pkg/backend/httpstate/backend.go

```diff
@@ -1129,7 +1131,64 @@ func (b *cloudBackend) Preview(ctx context.Context, stack backend.Stack,
 func (b *cloudBackend) Update(ctx context.Context, stack backend.Stack,
 	op backend.UpdateOperation,
 ) (sdkDisplay.ResourceChanges, error) {
-	return backend.PreviewThenPromptThenExecute(ctx, apitype.UpdateUpdate, stack, op, b.apply)
+	return backend.PreviewThenPromptThenExecute(ctx, apitype.UpdateUpdate, stack, op, b.apply, b)
+}
+
+func (b *cloudBackend) isCopilotFeatureEnabled(projectName tokens.PackageName, opts display.Options) bool {
+	// Have copilot features been requested by specifying the --copilot flag to the cli
+	if !opts.ShowCopilotFeatures {
+		return false
+	}
+
+	// Is copilot enabled this project in Pulumi Cloud
+	enabled, ok := b.copilotEnabledForProject[projectName]
+	contract.Assertf(ok,
+		"copilotEnabledForProject has not been set for project %q. only available after an update has been started.",
+		projectName)
+	return enabled
+}
+
+// explain takes engine events, renders them out to a buffer as something similar to what the user sees
+// in the CLI, and then explains the output with Copilot.
+func (b *cloudBackend) explain(
+	stackRef backend.StackReference,
+	op backend.UpdateOperation,
+	events []engine.Event,
+	opts display.Options,
+) (string, error) {
+	renderer := display.NewCaptureProgressEvents(
+		stackRef.Name(),
+		op.Proj.Name,
+		display.Options{},
+		true,
+		apitype.UpdateUpdate,
+	)
+	renderer.ProcessEventSlice(events)
+	output := renderer.Output()
+
+	if len(output) == 0 {
+		return "", errors.New("no output from preview")
+	}
+
+	stackID, err := b.getCloudStackIdentifier(stackRef)
+	if err != nil {
+		return "", err
+	}
+	orgName := stackID.Owner
+
+	display.RenderCopilotThinking(opts)
+	summary, err := b.client.ExplainPreviewWithCopilot(context.Background(), orgName, output)
```

Comment:
Don't we want to pass the real context down here to be able to cancel the request on Ctrl-C?

Reply:
Yeah good call